### PR TITLE
Restore ability to compile with C++

### PR DESCRIPTION
**Pull request by kgaillot**

### CODE REVIEW

**Review by nrwahl2 — COMMENTED** [review]
You were right, that was annoying.

File: include/crm/cluster/internal.h [hunk 1]
```diff
@@ -8,14 +8,17 @@
...
 
 #ifndef PCMK__CRM_CLUSTER_INTERNAL__H
-#  define PCMK__CRM_CLUSTER_INTERNAL__H
+#define PCMK__CRM_CLUSTER_INTERNAL__H
 
-#  include <stdbool.h>
-#  include <stdint.h>       // uint32_t, uint64_t
+#include <stdbool.h>
+#include <stdint.h>         // uint32_t, uint64_t
+#include <glib.h>           // gboolean
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply include/crm/cluster/internal.h:16]: Yeah that's a decent approach. (Of course in our case crm_internal.h always has to be first if used.)

File: doc/sphinx/Pacemaker_Development/c.rst [hunk 2]
```diff
@@ -271,39 +271,39 @@ generally one of "GNU General Public License version 2 or later (GPLv2+)"
 or "GNU Lesser General Public License version 2.1 or later (LGPLv2.1+)".
 
 Header files should additionally protect against multiple inclusion by defining
-a unique symbol of the form ``PCMK__<capitalized_header_name>__H``. For
-example:
+a unique symbol of the form ``PCMK__<capitalized_header_name>__H``., and
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply doc/sphinx/Pacemaker_Development/c.rst:274]: thx

File: lib/pacemaker/pcmk_sched_clone.c [hunk 3]
```diff
@@ -74,14 +74,14 @@ pcmk__clone_assign(pcmk_resource_t *rsc, const pcmk_node_t *prefer,
...
+    pe__show_node_scores(!pcmk_is_set(rsc->priv->scheduler->flags,
                                       pcmk__sched_output_scores),
-                         rsc, __func__, rsc->private->allowed_nodes,
-                         rsc->private->scheduler);
+                         rsc, __func__, rsc->priv->allowed_nodes,
+                         rsc->priv->scheduler);
 
-    rsc->private->children = g_list_sort(rsc->private->children,
+    rsc->priv->children = g_list_sort(rsc->priv->children,
                                          pcmk__cmp_instance);
```
[the remark this answers is not in the document]
Comment (kgaillot) [reply lib/pacemaker/pcmk_sched_clone.c:83]: Can't believe how many of these I missed ... I fixed a ton

**Comment by kgaillot** [comment]
updated to address review and issue found by ci

File: include/crm/cluster/internal.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (nrwahl2) [reply include/crm/cluster/internal.h:17]: Lol you didn't have to put it back, was just letting you know what I'd started doing. That works though. And yeah acknowledged about `crm_internal.h`

**Review by nrwahl2 — COMMENTED** [review]
I'm good with the changes but CI failed on Debian-based systems from test-headers:
```
14:02:55                   from ../include/crm/common/mainloop.h:81,
14:02:55                   from /tmp/test-headers-GEHMqzXgpO.c:1:
14:02:55  /usr/include/c++/12/bits/functexcept.h:101:3: error: conflicting declaration of C function ‘void std::__throw_ios_failure(const char*, int)’
14:02:55    101 |   __throw_ios_failure(const char*, int) __attribute__((__noreturn__));
14:02:55        |   ^~~~~~~~~~~~~~~~~~~
14:02:55  /usr/include/c++/12/bits/functexcept.h:98:3: note: previous declaration ‘void std::__throw_ios_failure(const char*)’
14:02:55     98 |   __throw_ios_failure(const char*) __attribute__((__noreturn__));
14:02:55        |   ^~~~~~~~~~~~~~~~~~~
14:02:55  In file included from /usr/include/c++/12/bits/stl_algobase.h:63:
14:02:55  /usr/include/c++/12/ext/numeric_traits.h:50:3: error: template with C linkage
14:02:55     50 |   template<typename _Tp>
14:02:55        |   ^~~~~~~~
14:02:55  ../include/crm/common/mainloop.h:18:1: note: ‘extern "C"’ linkage started here
14:02:55     18 | extern "C" {
14:02:55        | ^~~~~~~~~~
14:02:55  /usr/include/c++/12/ext/numeric_traits.h:60:3: error: template with C linkage
14:02:55     60 |   template<typename _Value>
14:02:55        |   ^~~~~~~~
14:02:55  ../include/crm/common/mainloop.h:18:1: note: ‘extern "C"’ linkage started here
14:02:55     18 | extern "C" {
14:02:55        | ^~~~~~~~~~
... many more ...
```

**Comment by kgaillot** [comment]
Added some commits on top to address CI issue and do more cleanup

**Comment by kgaillot** [comment]
updated test-headers commit to use bsd-compatible sed expression. fedora 40 looked like a timeout, not sure if it's a real issue

**Review by nrwahl2 — APPROVED** [review]

**Pull request merged**